core/src/test/java/kafka/test/junit/ClusterTestExtensions.java

```diff
@@ -119,7 +138,21 @@ public Stream<TestTemplateInvocationContext> provideTestTemplateInvocationContex
         return generatedContexts.stream();
     }
 
+    @Override
+    public void beforeEach(ExtensionContext context) {
+        detectThreadLeak = DetectThreadLeak.of(thread ->
+                SKIPPED_THREAD_PREFIX.stream().noneMatch(prefix -> thread.getName().startsWith(prefix)));
+    }
 
+    @Override
+    public void afterEach(ExtensionContext context) throws InterruptedException {
+        if (detectThreadLeak == null) {
+            return;
+        }
+        TestUtils.waitForCondition(() -> detectThreadLeak.newThreads().isEmpty(),
```

Comment:
I think this waitForCondition will cover up certain thread leaks where the test started shutdown of the thread, but didn't call join() to ensure the thread was completely stopped. I think these are leaks that would be good to find and fix.

Reply:
> I think this waitForCondition will cover up certain thread leaks where the test started shutdown of the thread, but didn't call join() to ensure the thread was completely stopped. I think these are leaks that would be good to find and fix.

that is a great idea, but I guess the related errors will spring up. For example, I apply the idea on my local, and Zk SessionTracker show up. We can fix them one by one to avoid random failed tests, but that will make this PR pending for a while. 

Maybe we can open a jira as follow-up to fix ignored "join()", and eventually we can remove the waiting from thread leak detection.

@gharris1727 WDYT?

Reply:
> that will make this PR pending for a while.

I took the strategy of fixing the leaks before enforcement with #14783 and it is true that PR has been open a long time at this point. I don't think that's necessarily a problem, as a ton of leaks have already been addressed.

Since this is a replacement for the existing verifyNoUnexpectedThreads that I just want gone asap, i'm fine with this as an intermediate.

Reply:
I open https://issues.apache.org/jira/browse/KAFKA-17151 as follow-up